cooked-validators/src/Cooked/Tx/Constraints/Type.hs

```diff
@@ -532,131 +721,184 @@ txSkelMintsValue =
           amount
     )
 
--- * Transaction inputs
+-- * Transaction outputs
 
-type SpendsScriptConstrs a =
-  ( Pl.ToData (Pl.RedeemerType a),
-    Pl.UnsafeFromData (Pl.DatumType a),
-    Show (Pl.DatumType a),
-    Show (Pl.RedeemerType a),
-    Pl.Eq (Pl.RedeemerType a),
-    Typeable a
-  )
+class IsTxSkelOutAllowedOwner a where
+  toPKHOrValidator :: a -> Either Pl.PubKeyHash (Pl.Versioned Pl.Validator)
 
-data TxSkelIn where
-  SpendsScript ::
-    SpendsScriptConstrs a =>
-    { spendingValidator :: Pl.TypedValidator a,
-      inputRedeemer :: Pl.RedeemerType a
-    } ->
-    TxSkelIn
-  SpendsPK :: TxSkelIn
+instance IsTxSkelOutAllowedOwner Pl.PubKeyHash where
+  toPKHOrValidator = Left
 
-deriving instance Show TxSkelIn
+instance IsTxSkelOutAllowedOwner (Pl.TypedValidator a) where
+  toPKHOrValidator = Right . Pl.vValidatorScript
 
-makeLensesFor
-  [ ("consumedOutput", "consumedOutputL"),
-    ("spendingValidator", "spendingValidatorAT"),
-    ("inputRedeemer", "inputRedeemerAT")
-  ]
-  ''TxSkelIn
+data TxSkelOut where
+  Pays ::
+    ( IsOutput o,
+      Show (OwnerType o),
+      IsTxSkelOutAllowedOwner (OwnerType o),
+      Typeable (OwnerType o),
+      ToCredential (OwnerType o),
+      Show (DatumType o),
+      ToOutputDatum (DatumType o),
+      Pl.ToData (DatumType o), -- If this seems redundant with the 'ToOutputDatum' constraint, see the [note on TxSkelOut data].
```

Comment:
I am wondering if all these type constraints should really be depicted here instead of somewhere else, like in the requirements for IsOutput.

Reply:
would there be other examples that would restrict it too much? I would advocate to put as much as convenient and possible into `IsOutput` directly, even if some other restrictions must be put in this function's signature, such as the one you mention that would be too restrictive for `TxOut`.